### PR TITLE
Replacing master with primary

### PR DESCRIPTION
**Pull request by nenox8885**
This is following from the work of https://github.com/zendesk/active_record_shards/pull/216, replace the methods on_master with on_primary

### CODE REVIEW

File: lib/active_record_shards/deprecation.rb
Comment (brianburnszd) [inline lib/active_record_shards/deprecation.rb:9]: @craig-day without setting this log behaviour, we found that deprecated methods were raising errors in test and development. happy to keep this for this gem?

ref:
> Deprecation warnings raised by gems are not affected by this setting because they happen before Rails boots up.
https://edgeapi.rubyonrails.org/classes/ActiveSupport/Deprecation/Behavior.html#method-i-behavior-3D

**Review by brianburnszd — APPROVED** [review]

File: lib/active_record_shards/connection_switcher.rb
Comment (mriddle) [inline lib/active_record_shards/connection_switcher.rb:117]: `on_cx_switch_block` and `on_primary_or_replica` are public methods, this is a breaking change. How about this instead?

```suggestion
      @disallow_replica += 1 if [:primary, :master].include?(which)
```

File: lib/active_record_shards/connection_switcher.rb
Comment (mriddle) [inline lib/active_record_shards/connection_switcher.rb:132]: `on_cx_switch_block` and `on_primary_or_replica` are public methods, this is a breaking change. How about this instead?

```suggestion
      @disallow_replica -= 1 if [:primary, :master].include?(which)
```

File: lib/active_record_shards/connection_switcher.rb
Comment (mriddle) [reply lib/active_record_shards/connection_switcher.rb:117]: We should add a deprecation warning too.

File: lib/active_record_shards/sql_comments.rb
Comment (mriddle) [inline lib/active_record_shards/sql_comments.rb:7]: This might deserve a mention in the changelog. Clients may have been using those comments to perform operations.

**Review by mriddle — CHANGES_REQUESTED** [review]
Nice work! Could you please add an entry to the changelog?

There was a missing reference [in the README](https://github.com/zendesk/active_record_shards/blob/argonauts/lab-day-2020-08-27/README.md#debugging).

> Show if a query went to ~master~primary or replica in the logs:

I could only spot one issue, comments below.

File: lib/active_record_shards/connection_switcher.rb
Comment (bquorning) [inline lib/active_record_shards/connection_switcher.rb:178]: This warning is not necessary, since any `:master` option would just be ignored anyway – it isn’t being used in this method.

But looking harder at this change made me realize that a bug was introduced for the `:replica` code – `current_shard_selection.on_replica` is _not_ being set when the option `:slave` is used instead of `:replica`! Changing tests and implementation at the same time _does_ have its downsides 😅

**Review by bquorning — COMMENTED** [review]
I found still _one_ mention of "master", in the Readme.

File: Changelog.md
Comment (bquorning) [inline Changelog.md:22]: Also, the method `on_master_or_replica` has been removed.

File: lib/active_record_shards/deprecation.rb
Comment (bquorning) [reply lib/active_record_shards/deprecation.rb:9]: Will it still be possible to configure it to use e.g. `:raise`?

**Review by bquorning — COMMENTED** [review]
I a little bit concerned that we change all the tests too, so there are no tests covering that the old behaviour is still working.

File: Changelog.md
Comment (bquorning) [inline Changelog.md:11]: Could you add Changelog.md to one of the .github/term-check files?

File: Changelog.md
Comment (brianburnszd) [reply Changelog.md:22]: the last PR and this will be one release, so `on_master_or_replica` method may never be released?

File: lib/active_record_shards/deprecation.rb
Comment (brianburnszd) [reply lib/active_record_shards/deprecation.rb:9]: removing this, I was incorrect, in a Rails app `config.active_support.deprecation = :log` works as expected.

File: lib/active_record_shards/configuration_parser.rb
Comment (brianburnszd) [inline lib/active_record_shards/configuration_parser.rb:32]: Rubocop is complaining bout not using `next` here, but I'm not sure if `next` makes sense in this case
```
lib/active_record_shards/configuration_parser.rb:32:9: C: Style/Next: Use next to skip iteration.
        if legacy_replica_conf = env_config.delete('slave')
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by brianburnszd** [comment]
@bquorning I added `test/configuration_parser_legacy_test.rb` and found a regression 🎉

File: lib/active_record_shards/configuration_parser.rb
Comment (bquorning) [inline lib/active_record_shards/configuration_parser.rb:35]: ❤️ Of course.

This 2nd `if` block was previously places above the 1st `if` block, and then of course it would have worked fine.

File: lib/active_record_shards/connection_switcher.rb
Comment (craig-day) [inline lib/active_record_shards/connection_switcher.rb:168]: why is this being removed?

File: lib/active_record_shards/shard_selection.rb
Comment (craig-day) [inline lib/active_record_shards/shard_selection.rb:88]: this isn't private, which is why I left the option in there. In theory, any external consumer could call this options method and then this would be an immediate breaking change.

File: test/configuration_parser_test.rb
Comment (craig-day) [inline test/configuration_parser_test.rb:14]: 💣

File: lib/active_record_shards/shard_selection.rb
Comment (bquorning) [reply lib/active_record_shards/shard_selection.rb:88]: I think the chances of anyone using `ShardSelection#options` directly is very slim. And it would be nice to not issue warnings because of ARS doing an internal call with deprecated terminology (which happens when calling `ConnectionSwitcher#switch_connection`)

Can you help me assess the risk of leaving this change as it is (with `:slave` key removed)? It might be enough if we just mention it as a breaking change in the (brand new) changelog.

Comment (brianburnszd) [reply lib/active_record_shards/shard_selection.rb:88]: @craig-day this is the issue @bquorning described, overall it seems best to remove it, and make a note in Changelog.

File: lib/active_record_shards/connection_switcher.rb
Comment (brianburnszd) [reply lib/active_record_shards/connection_switcher.rb:168]: this is removed because we're changing options in ShardSelection, which is what gets passed in here:

```
def options
      { shard: @shard, replica: @on_replica } # no more :slave option
end
```

File: lib/active_record_shards/shard_selection.rb
Comment (brianburnszd) [reply lib/active_record_shards/shard_selection.rb:88]: this allows removing the check / deprecation warning raised internally in `connection_switcher` (https://github.com/zendesk/active_record_shards/pull/237#discussion_r533829639)

Comment (craig-day) [reply lib/active_record_shards/shard_selection.rb:88]: I agree that the risk is rather low for removing this, was just calling out that technically it is a breaking change. If we're OK with documenting this and removing it before we bump the next major version, I think that is reasonable.

File: lib/active_record_shards/connection_switcher.rb
Comment (craig-day) [reply lib/active_record_shards/connection_switcher.rb:168]: I suppose if we proceed with removing it below, then this makes sense.

**Review by craig-day — APPROVED** [review]
I think this looks good! Let's wait for @bquorning or @mriddle to also sign off as the database gem experts.

File: lib/active_record_shards/shard_selection.rb
Comment (brianburnszd) [reply lib/active_record_shards/shard_selection.rb:88]: added this change to changelog

**Review by mriddle — APPROVED** [review]

**Comment by mriddle** [comment]
Please squash the commits before merging 🙏 💖

**Pull request merged**
